server/src/services/metadata.service.ts

```diff
@@ -487,6 +496,124 @@ export class MetadataService implements OnEvents {
     }
   }
 
+  async searchPerson(ownerId: AssetEntity['ownerId'], dto: SearchPeopleDto): Promise<PersonResponseDto[]> {
+    return this.personRepository.getByName(ownerId, dto.name, { withHidden: dto.withHidden });
+  }
+
+  private async applyTaggedFaces(asset: AssetEntity, tags: ImmichTags) {
+    this.logger.debug(`Starting faces extraction from tags (${asset.id})`);
+
+    const faceSet = new Set();
+
+    if (!tags.RegionInfo) {
+      return true;
+    }
+
+    interface Region {
+      imageWidth: number;
+      imageHeight: number;
+      boundingBoxX1: number;
+      boundingBoxX2: number;
+      boundingBoxY1: number;
+      boundingBoxY2: number;
+    }
+
+    interface MetadataFaceResult {
+      Name?: string;
+      Type?: string;
+      Region: Region;
+    }
+
+    function toPixels(region: Region, unit: string): Region {
+      const pixelRegion: Region = region;
+      if (unit.toLowerCase() == 'normalized') {
+        pixelRegion.boundingBoxX1 = Math.floor(region.boundingBoxX1 * region.imageWidth);
+        pixelRegion.boundingBoxX2 = Math.floor(region.boundingBoxX2 * region.imageWidth);
+        pixelRegion.boundingBoxY1 = Math.floor(region.boundingBoxY1 * region.imageHeight);
+        pixelRegion.boundingBoxY2 = Math.floor(region.boundingBoxY2 * region.imageHeight);
+      }
+      return pixelRegion;
+    }
+
+    for (const region of tags.RegionInfo?.RegionList) {
+      const faceResult: MetadataFaceResult = {
+        Name: region.Name,
+        Type: region.Type,
+        Region: toPixels(
+          {
+            imageWidth: tags.RegionInfo?.AppliedToDimensions?.W,
+            imageHeight: tags.RegionInfo?.AppliedToDimensions?.H,
+            boundingBoxX1: region.Area.X - region.Area.W / 2,
+            boundingBoxY1: region.Area.Y - region.Area.H / 2,
+            boundingBoxX2: region.Area.X + region.Area.W / 2,
+            boundingBoxY2: region.Area.Y + region.Area.H / 2,
+          },
+          region.Area.Unit,
+        ),
+      };
+
+      faceSet.add(faceResult);
+    }
+
+    const faces = [...faceSet] as MetadataFaceResult[];
+
+    this.logger.debug(`${faces.length} faces detected in ${asset.originalPath}`);
+
+    // cleanup faces imported from metadata so we can update them if we're re-reading to avoid duplications
+    await this.personRepository.deleteFaceFromAsset(asset.id, SourceType.EXIF);
+
+    for (const face of faces) {
```

Comment:
This part has a lot of duplication with the face detection handler. You should be able to extract this out into a separate method (or methods).

Reply:
Unsure on how should I proceed here. Would it be ok to split the detection handler in smaller pieces so that I can reuse some in the metadata processing avoiding such duplication?

Reply:
So looking at this again, this logic involves using repositories and has some key differences in how it works, so might be best to leave it as-is. "Duplication is better than the wrong abstraction", and all that.